enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,283 @@
+	// Example: "https://expected.OIDC.issuer/"
+	OIDCIssuer string `json:"oidcIssuer,omitempty"`
+
+	// SubjectEmail holds the email address of the subject.
+	// Example: "expected-signing-user@example.com"
+	// Requires OIDCIssuer to be set and KeyData to be empty.
+	SubjectEmail string `json:"subjectEmail,omitempty"`
+
+	// KeyData contains inline base64 data of the public key.
+	// Can be empty if the image got signed keyless.
+	// Requires OIDCIssuer and SubjectEmail to be empty.
+	KeyData string `json:"keyData,omitempty"`
+
+	// SignedIdentity specifies what image identity the signature
+	// claims about the image.
+	SignedIdentity Identity `json:"signedIdentity,omitempty"`
+}
+
+type Identity struct {
+	// Defaults to "matchRepository"
+	IdentityMatchPolicy IdentityMatchPolicy `json:"identityMatchPolicy,…"`
+	DockerRepository    string              `json:"dockerRepository,omitempty"`
+}
+
+type IdentityMatchPolicy string
+
+const (
+	IdentityMatchPolicyMatchRepository = "matchRepository"
+	IdentityMatchPolicyExactRepository = "exactRepository"
+)
+
+type ImagePolicyStatus struct {
+	// PolicyJSON contains the whole policy applied to the namespace
+	// which got written to disk. This includes cluster-wide policies
+	// from the `openshift-config` namespace as well.
+	PolicyJSON string `json:"policyJSON,omitempty"`
+
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Update container runtime config controller to watch ImagePolicy
+
+MCO container runtime config controller ImagePolicy syncHandler updates the signature verification configurations:
+- retrieves all the ImagePolicy instances on the cluster.
+- validates conflicting data from ImagePolicy instances.
+- adds following configurations to a single machine config
+  - merges the instances for each namespace into a single [containers-policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) in a predefined /path/to/policies/\<NAMESPACE\>.json.  
+  If no policy exists in a namespace, use the current default of:
+
+    ```yaml
+    { "default": [{ "type": "insecureAcceptAnything" }] }
+    ```
+
+    If a single policy exists, use the default:
+    
+    ```yaml
+    { "default": [{"type": "reject"}] }
+    ```
+
+  - adds [/etc/containers/registries.d/*.yaml](https://github.com/containers/image/blob/main/docs/containers-registries.d.5.md) to allow matching sigstore signatures, for example:
+
+    ```yaml
+    docker:
+      my-registry/image:
+        use-sigstore-attachments: true
+    ```
+
+    Alternatively, we can enable it node wide:
+
+    ```yaml
+    default-docker:
+      use-sigstore-attachments: true
+    ```
+- merges cluster scoped with namespaced ImagePolicy as a json string and update `status` of each ImagePolicy.
+Once an ImagePolicy object gets created/updated, container runtime config controller will call the syncHandler.
```

Comment:
An opportunity here is to make the sync do an overall sync of all ImagePolicy objects whenever this happens (as opposed to just the object being updated), to decrease the chances of something mis-syncing across namespaces, and to align it better for the MCO

Reply:
yes, the syncHandler would do the sync for all ImagePolicy objects, and the json string of the merged sigstore policies will be written to the `status` of all existing ImagePolicy objects.